src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1685,8 +1685,21 @@ internal bool CheckValueKind(BoundExpression expr, BindValueKind valueKind, Diag
                 case BoundKind.EventAccess:
                     return CheckEventValueKind((BoundEventAccess)expr, valueKind, diagnostics);
                 case BoundKind.DynamicMemberAccess:
+                    {
+                        // dynamic access needs to check if the source of the
+                        // dynamic object is from, for example, a set-only property
+                        var access = (BoundDynamicMemberAccess)expr;
+                        return CheckValueKind(access.Receiver, BindValueKind.RValue, diagnostics);
```

Comment:
It feels like this check is too late. I would expect it to be performed before BoundDynamicMemberAccess is created. Here we are supposed to check BoundDynamicMemberAccess itself